debezium-core/src/main/java/io/debezium/connector/common/OffsetReader.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.common;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.connect.storage.OffsetStorageReader;
+
+import io.debezium.pipeline.spi.OffsetContext;
+
+/**
+ * Provides access to the partition offsets stored by connectors.
+ */
+public class OffsetReader<P extends Partition, O extends OffsetContext, L extends OffsetContext.Loader<O>> {
+
+    private final OffsetStorageReader reader;
+    private final L loader;
+
+    public OffsetReader(OffsetStorageReader reader, L loader) {
+        this.reader = reader;
+        this.loader = loader;
+    }
+
+    /**
+     * Given the collection of connector-specific task partitions, returns their respective connector-specific offsets.
+     *
+     * If there is no offset stored for a given partition, the corresponding key will be mapped to a null.
+     */
+    public Map<P, O> offsets(Collection<P> partitions) {
+        Collection<Map<String, String>> sourcePartitions = partitions.stream()
+                .map(Partition::getSourcePartition)
+                .collect(Collectors.toCollection(HashSet::new));
+
+        Map<Map<String, String>, Map<String, Object>> sourceOffsets = reader.offsets(sourcePartitions);
```

Comment:
`Map<Map<String, String>, Map<String, Object>>` uff, that's tough. How about we abstract this in our own representation; i.e. instead of returning `Map<P, O>`, return a dedicated `Offsets` contract which offers the right API for the task?

Reply:
> `Map<Map<String, String>, Map<String, Object>>` uff, that's tough.

This is just what `org.apache.kafka.connect.storage.OffsetStorageReader#offsets` returns. Why is it a problem given that it's just a local variable?

> How about we abstract this in our own representation; i.e. instead of returning Map<P, O>, return a dedicated Offsets contract which offers the right API for the task?

This is how it's currently [implemented](https://github.com/sugarcrm/debezium/blob/1ea6f6cefffe1ab95deb87f553e494a5072c5da0/debezium-core/src/main/java/io/debezium/connector/common/TaskOffsetContext.java#L45-L64) in our feature branch.

The thing is that this object won't have any logic, but its internal `Map<P, O>` map needs to be iterated and accessed in other ways throughout the codebase. So every consumer will end up calling something like `Offsets#toMap()` and using the internal `Map<P, O>`.

This is again where I followed the Kafka Connect example and didn't introduce an unnecessary wrapper. Do you believe we should introduce it?

Reply:
> Why is it a problem given that it's just a local variable?

Ah, it's just whenever I see such generic structure of maps of maps, I got the strong urge to introduce some more expressive domain-specific type. Of course we're bound here to what's exposed in the Kafka API.

> will end up just calling something like Offsets#toMap()

Ok, when all clients/callers will convert this into a map anyways, there's indeed not much benefit to changing this. Leaving it up to you to decide.